### ISSUE TITLE
Debug timeouts with some tests in CI

### ISSUE THREAD

**Issue by kbendick** [apache/iceberg#3091]
Sometimes, the Java test suites timeout.

It appears as tough they're trying to start the Spark 3 tests, per the logs. We can add further debug logs to investigate, but it requires setting a repository secret in GitHub (either via the REST api or via the CLI or otherwise... CLI seems easiest).

**Comment by kbendick** [comment]
To enable debugging logging is documented here: https://docs.github.com/en/actions/monitoring-and-troubleshooting-workflows/enabling-debug-logging

I'll see if I have the permissions for this and try it out.

**Comment by kbendick** [comment]
There are two types of debug logging. Both are enabled by setting a GitHub secret in the repo to `true`.

Enable runner diagnostic logging (information about how the runner processed and set up jobs etc): set the repo secret `ACTIONS_RUNNER_DEBUG` to `true`.

Increasing log verbosity - Set the repo secret `ACTIONS_STEP_DEBUG` to `true`.

Encrypted secrets (includes `gh` cli): https://docs.github.com/en/actions/reference/encrypted-secrets (basically `gh secrets create ....`).

**Comment by kbendick** [comment]
Link to the discussion in Slack: https://apache-iceberg.slack.com/archives/C025PH0G1D4/p1630603003176300

**Comment by kbendick** [comment]
TLDR - If you can `gh secret list` from the repo, please reach out to me!

**Comment by kbendick** [comment]
The command for setting the config for just this repo would be `gh secret set ACTIONS_STEP_DEBUG -btrue --repos="iceberg" --org="apache" --visibility=all`

**Comment by kbendick** [comment]
For the other secret, the command would be `gh secret set ACTIONS_RUNNER_DEBUG -btrue --repos="iceberg" --org="apache" --visibility=all`

**Comment by kbendick** [comment]
First I'm going to just enable debug logging from gradle's point of view to confirm.

**Comment by kbendick** [comment]
So far there's only been one time out, since merging the added logs, which was actually from the merge commit of that PR.

We should probably look into whether or not we want CI to run AGAIN after merge (e.g. when Ryan merged the above PR that had already passed CI, CI ran again).

Here's a link to that: https://github.com/apache/iceberg/actions/runs/1222841491

The Java 8 test suite is the one that timed out. Seems to have been working on a Parquet writer test suite.

The test suite running seemed to have a relatively large number of tasks (at least 120), but I'm not sure if that was just for that stage or if that was aggregated with other stages in other tests. Will analyze the logs more closely now that I've downloaded them.

I have a raw zip of the logs from that test if anybody wants to see them as well.

**Comment by kbendick** [comment]
After enabling the info level logging from gradle, a timeout was reported by @nastra for the following job:

https://github.com/apache/iceberg/pull/3024/checks

The logs seem to be spewing the following Flink related logs repeatedly until the 6h point is reached.

I am downloading these logs as well, but it seems as though this particular Flink test shuts down / enters the `FINISHED` state before it can complete an active checkpoint, and is retrying that constantly.

I believe there is a configuration parameter for maximum number of allowed failed commits. I will try adding that and see if that helps.

```
2021-09-13T08:19:47.7896411Z > Task :iceberg-flink:test
2021-09-13T08:19:47.7899950Z     [Checkpoint Timer] INFO org.apache.flink.runtime.checkpoint.CheckpointCoordinator - Checkpoint triggering task Source: rightCustomSource -> rightIcebergSink-rightIcebergSink -> rightIcebergSink-IcebergStreamWriter (1/1) of job 437e46445e777ca2231677f60f87496a is not in state RUNNING but FINISHED instead. Aborting checkpoint.
2021-09-13T08:19:47.7905489Z     [Checkpoint Timer] INFO org.apache.flink.runtime.checkpoint.CheckpointCoordinator - Checkpoint triggering task Source: rightCustomSource -> rightIcebergSink-rightIcebergSink -> rightIcebergSink-IcebergStreamWriter (1/1) of job 437e46445e777ca2231677f60f87496a is not in state RUNNING but FINISHED instead. Aborting checkpoint.
2021-09-13T08:19:47.7914766Z     [Checkpoint Timer] INFO org.apache.flink.runtime.checkpoint.CheckpointCoordinator - Checkpoint triggering task Source: rightCustomSource -> rightIcebergSink-rightIcebergSink -> rightIcebergSink-IcebergStreamWriter (1/1) of job 437e46445e777ca2231677f60f87496a is not in state RUNNING but FINISHED instead. Aborting checkpoint.
2021-09-13T08:19:47.7920502Z     [Checkpoint Timer] INFO org.apache.flink.runtime.checkpoint.CheckpointCoordinator - Checkpoint triggering task Source: rightCustomSource -> rightIcebergSink-rightIcebergSink -> rightIcebergSink-IcebergStreamWriter (1/1) of job 437e46445e777ca2231677f60f87496a is not in state RUNNING but FINISHED instead. Aborting checkpoint.
```

**Comment by kbendick** [comment]
This stack overflow question might be of use for debugging this issue: https://stackoverflow.com/questions/65236052/flink-checkpoint-is-getting-aborted


```You can check your parallelism Settings, if your program's parallelism is greater than the source's parallelism, then some task will show completion because it doesn't have the data, and that would abort a checkpoint.```

**Comment by kbendick** [comment]
There's more discussion here: https://issues.apache.org/jira/browse/FLINK-2491

It seems as though the source finishes, so its tasks shut down, while other parts are still checkpointing, so it doesn't complete.

I think we'll need to update `BoundedTestSource` and its checkpoint abort logic.

**Comment by kbendick** [comment]
The CountdownLatch used here to keep the source open seems to be one possible solution: https://github.com/apache/flink/blob/313e20e8e03953a5e1cec9daa467f561ccfbd599/flink-connectors/flink-connector-files/src/test/java/org/apache/flink/connector/file/sink/StreamingExecutionFileSinkITCase.java#L55-L73

I notice that we have some handling for this case, so I think it's just the addition of this two separate DAGs test that is causing this problem.

**Comment by kbendick** [comment]
This can be closed as the actual problematic test is resolved. We should make an issue for the test that was closed though so leaving this up until I do that.

**Comment by kbendick** [comment]
The test that was closed is a Flink issue. It is resolved as of Flink 1.14 (batch jobs take a final checkpoint so that operators don't necessarily close early).
